In [1]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier

In [2]:
iris = load_iris()
x = iris.data[:,2:]
y = iris.target


In [3]:
tree_clf = DecisionTreeClassifier(max_depth=2)
tree_clf.fit(x,y)

DecisionTreeClassifier(max_depth=2)

**to visualize a Decision Tree**

In [4]:
from sklearn.tree import export_graphviz

In [5]:
export_graphviz(
    tree_clf,
    out_file=("iris_tree.dot"),
    feature_names= iris.feature_names[2:],
    class_names= iris.target_names,
    rounded = True,
    filled = True
)

estimating class probabities
- let's say petal length = 5 & petal_width = 1.5

In [6]:
tree_clf.predict_proba([[5,1.5]])

array([[0.        , 0.90740741, 0.09259259]])

0% for setosa
90.7% for vericolor
0.09% for virginica

**Regression with DT**

In [7]:
from sklearn.tree import DecisionTreeRegressor
tree_reg = DecisionTreeRegressor(max_depth=2)
tree_reg.fit(x,y)

DecisionTreeRegressor(max_depth=2)

In [8]:
export_graphviz(
    tree_reg,
    out_file=("iris_tree_reg.dot"),
    feature_names= iris.feature_names[2:],
    class_names= iris.target_names,
    rounded = True,
    filled = True
)

In [9]:
import numpy as np
import pandas as pd

In [10]:
from sklearn.datasets import make_moons

In [11]:
moons = make_moons(n_samples =10000, noise=0.4)
moons

(array([[ 0.36173243,  0.38765163],
        [-1.21990909,  0.78618793],
        [ 0.25330711, -0.68937462],
        ...,
        [ 0.21909657,  1.09244758],
        [-0.30217231,  0.39107475],
        [ 2.47856104, -0.33211952]]),
 array([1, 0, 1, ..., 0, 0, 1], dtype=int64))

In [12]:
from sklearn.model_selection import train_test_split

x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.2,random_state=42)

In [13]:
from sklearn.model_selection import GridSearchCV

In [16]:
param_grid = [{'criterion':['gini','entropy'],
              'max_depth':np.arange(1,20),
              'max_leaf_nodes':np.arange(1,20),
              }]

In [22]:
grid = GridSearchCV(DecisionTreeClassifier(random_state=42),param_grid,cv=5)

In [28]:
%%time
grid.fit(x_train,y_train)

Wall time: 15.6 s


C:\Users\Oluwajuwon\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:372: FitFailedWarning: 
190 fits failed out of a total of 3610.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
190 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\Oluwajuwon\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 680, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\Oluwajuwon\anaconda3\lib\site-packages\sklearn\tree\_classes.py", line 937, in fit
    super().fit(
  File "C:\Users\Oluwajuwon\anaconda3\lib\site-packages\sklearn\tree\_classes.py", line 314, in fit
    raise ValueError(
ValueError: max_leaf_nodes 1 must be either None or large

GridSearchCV(cv=5, estimator=DecisionTreeClassifier(random_state=42),
             param_grid=[{'criterion': ['gini', 'entropy'],
                          'max_depth': array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19]),
                          'max_leaf_nodes': array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19])}])

In [29]:
grid.best_estimator_

DecisionTreeClassifier(max_depth=3, max_leaf_nodes=4, random_state=42)

In [30]:
tree = DecisionTreeClassifier(criterion='entropy',max_depth=3,max_leaf_nodes=4,random_state=42)

In [31]:
tree.fit(x_train,y_train)

DecisionTreeClassifier(criterion='entropy', max_depth=3, max_leaf_nodes=4,
                       random_state=42)

In [32]:
from sklearn.metrics import accuracy_score

In [33]:
pred = tree.predict(x_test)

In [34]:
accuracy_score(pred,y_test)

1.0

In [35]:
'a'

'a'